brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+    lazy val features = instance.features.mapValues { value => splitter.create(value, instance.target) }
+
+    for
+      ((treeIndex, tree) <- trees.toList;
+      i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+      leaf <- tree.leafFor(instance.features).toList if stopper.shouldSplit(leaf.target);
+      (feature, stats) <- features if (sampler.includeFeature(feature, treeIndex, leaf.index)))
+        yield (treeIndex, leaf.index, feature) -> stats
+  }
+
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K, v1: V1) = {
+    tree.leafAt(leafIndex).toList.flatMap { leaf =>
+      splitter
+        .split(leaf.target, v1)
+        .map { rawSplit =>
+          val (split, goodness) = evaluator.evaluate(rawSplit)
+          (key, split, goodness)
         }
+    }
+  }
 
-        tree.updateByLeafIndex(newNodes.lift)
+  def update(tree: Tree[K,V,T], map: Map[Int,V2]) = {
+    tree.growByLeafIndex { index =>
+      for (
+        (feature, split, _) <- map.get(index).toList;
+        (predicate, target) <- split.predicates
+      ) yield (feature, predicate, target, ())
     }
   }
 
-  def updateTargets: Trainer[K, V, T] =
-    updateLeaves {
-      case (treeIndex, leaf, instances) =>
-        val target = implicitly[Monoid[T]].sum(instances.map { _.target })
-        leaf.copy(target = target)
+  val semigroup1 = splitter.semigroup
+  val semigroup2 = new Semigroup[V2] {
+    def plus(a: V2, b: V2) = {
+      if (a._3 > b._3) a else b
     }
+  }
+}
+
 
-  def expand(times: Int)(implicit splitter: Splitter[V, T], evaluator: Evaluator[V, T], stopper: Stopper[T]): Trainer[K, V, T] =
-    updateLeaves {
-      case (treeIndex, leaf, instances) =>
-        Tree.expand(times, treeIndex, leaf, splitter, evaluator, stopper, sampler, instances)
+trait OutputStep[K,V,T,O] {
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]): Seq[O]
+  def semigroup: Semigroup[O]
+}
+
+case class Validate[K,V,T:Semigroup,P,E](sampler: Sampler[K], error: Error[T, P,E], voter: Voter[T, P])
+  extends OutputStep[K,V,T,E] {
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+    val predictions =
+      for (
+        (treeIndex, tree) <- trees
+          if sampler.includeInValidationSet(instance.id, instance.timestamp, treeIndex);
+        target <- tree.targetFor(instance.features).toList
+      ) yield target
+
+    List(error.create(instance.target, voter.combine(predictions)))
+  }
+
+  val semigroup = error.semigroup
+}
+
+case class Trainer[K: Ordering, V, T: Monoid](
+    trainingData: Iterable[Instance[K, V, T]],
+    sampler: Sampler[K],
+    trees: List[Tree[K, V, T]]) {
+
+  def trainingStep(step: TrainingStep[K,V,T]): Trainer[K,V,T] = {
+    val treeMap = trees.zipWithIndex.map{case (t,i) => i->t}.toMap
+    var sums1 = Map[(Int,Int,step.K1),step.V1]()
+
+    trainingData.foreach{instance =>
```

Comment:
A bit of golf, but I think this should do roughly the same thing (including laziness):

```scala
val sums1 = MapAlgebra.rollupSum(trainingData.iterator.flatMap(step.prepare(treeMap, _)))
```